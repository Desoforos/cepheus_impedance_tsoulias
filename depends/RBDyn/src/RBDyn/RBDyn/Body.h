/*
 * Copyright 2012-2019 CNRS-UM LIRMM, CNRS-AIST JRL
 */

#pragma once

// includes
// std
#include <string>

// SpaceVecAlg
#include <SpaceVecAlg/SpaceVecAlg>

namespace rbd
{

/** Body representation.
 * Hold body id, name and spatial rigid body inertia of one body.
 */
class Body
{
public:
  Body() {}

  /**
   * @param rbInertia Body spatial rigid body inertia.
   * @param name Body name, must be unique in a multibody.
   */
  Body(const sva::RBInertiad & rbInertia, std::string name) : inertia_(rbInertia), name_(std::move(name)) {}

  /**
   * @param mass Body mass.
   * @param com Body center of mass.
   * @param inertia Body inertia matrix at body origin.
   * @param name Body name, must be unique in a multibody.
   */
  Body(double mass, const Eigen::Vector3d & com, const Eigen::Matrix3d & inertia, std::string name)
  : inertia_(mass, mass * com, inertia), name_(std::move(name))
  {
  }

  /// @return Body name.
  const std::string & name() const
  {
    return name_;
  }

  /// @return Body spatial rigid body inertia.
  const sva::RBInertiad & inertia() const
  {
    return inertia_;
  }

  bool operator==(const Body & b) const
  {
    return name_ == b.name_;
  }

  bool operator!=(const Body & b) const
  {
    return name_ != b.name_;
  }

private:
  sva::RBInertiad inertia_;
  std::string name_;
};

inline std::ostream & operator<<(std::ostream & out, const Body & b)
{
  out << "Body: " << b.name();
  return out;
}

} // namespace rbd
