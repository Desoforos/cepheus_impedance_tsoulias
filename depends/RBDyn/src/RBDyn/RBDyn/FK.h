/*
 * Copyright 2012-2019 CNRS-UM LIRMM, CNRS-AIST JRL
 */

#pragma once

#include <rbdyn/config.hh>

namespace rbd
{
class MultiBody;
struct MultiBodyConfig;

/**
 * Compute the forward kinematic of a MultiBody.
 * @param mb MultiBody used has model.
 * @param mbc Use q generalized position vector. Fill bodyPosW, jointConfig, motionSubspace
 * and parentToSon.
 */
RBDYN_DLLAPI void forwardKinematics(const MultiBody & mb, MultiBodyConfig & mbc);

/**
 * Safe version.
 * @see forwardKinematics.
 * @throw std::domain_error If there is a mismatch between mb and mbc.
 */
RBDYN_DLLAPI void sForwardKinematics(const MultiBody & mb, MultiBodyConfig & mbc);

} // namespace rbd
