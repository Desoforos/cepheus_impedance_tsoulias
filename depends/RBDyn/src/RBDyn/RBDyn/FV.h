/*
 * Copyright 2012-2019 CNRS-UM LIRMM, CNRS-AIST JRL
 */

#pragma once

#include <rbdyn/config.hh>

namespace rbd
{
class MultiBody;
struct MultiBodyConfig;

/**
 * Compute the forward velocity of a MultiBody.
 * @param mb MultiBody used has model.
 * @param mbc Use alpha generalized velocity vector, bodyPosW, jointConfig
 * and parentToSon.
 * Fill jointVelocity, bodyVelW and bodyVelB.
 */
RBDYN_DLLAPI void forwardVelocity(const MultiBody & mb, MultiBodyConfig & mbc);

/**
 * Safe version.
 * @see forwardVelocity.
 * @throw std::domain_error If there is a mismatch between mb and mbc.
 */
RBDYN_DLLAPI void sForwardVelocity(const MultiBody & mb, MultiBodyConfig & mbc);

} // namespace rbd
