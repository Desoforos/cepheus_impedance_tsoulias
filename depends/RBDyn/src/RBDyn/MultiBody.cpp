/*
 * Copyright 2012-2019 CNRS-UM LIRMM, CNRS-AIST JRL
 */

// associated header
#include "RBDyn/MultiBody.h"

// includes
// RBDyn
#include "RBDyn/Body.h"
#include "RBDyn/Joint.h"

namespace rbd
{

MultiBody::MultiBody() : nrParams_(0), nrDof_(0) {}

MultiBody::MultiBody(std::vector<Body> bodies,
                     std::vector<Joint> joints,
                     std::vector<int> pred,
                     std::vector<int> succ,
                     std::vector<int> parent,
                     std::vector<sva::PTransformd> Xto)
: bodies_(std::move(bodies)), joints_(std::move(joints)), pred_(std::move(pred)), succ_(std::move(succ)),
  parent_(std::move(parent)), Xt_(std::move(Xto)), jointPosInParam_(joints_.size()), jointPosInDof_(joints_.size()),
  nrParams_(0), nrDof_(0)
{
  for(int i = 0; i < static_cast<int>(bodies_.size()); ++i)
  {
    const auto ui = static_cast<size_t>(i);

    bodyNameToInd_[bodies_[ui].name()] = i;
    jointNameToInd_[joints_[ui].name()] = i;

    jointPosInParam_[ui] = nrParams_;
    jointPosInDof_[ui] = nrDof_;

    nrParams_ += joints_[ui].params();
    nrDof_ += joints_[ui].dof();
  }
}

} // namespace rbd
