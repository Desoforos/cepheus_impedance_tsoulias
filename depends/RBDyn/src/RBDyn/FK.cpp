/*
 * Copyright 2012-2019 CNRS-UM LIRMM, CNRS-AIST JRL
 */

// associated header
#include "RBDyn/FK.h"

// includes
// RBdyn
#include "RBDyn/Joint.h"
#include "RBDyn/MultiBody.h"
#include "RBDyn/MultiBodyConfig.h"

namespace rbd
{

void forwardKinematics(const MultiBody & mb, MultiBodyConfig & mbc)
{
  const std::vector<Joint> & joints = mb.joints();
  const std::vector<int> & pred = mb.predecessors();
  const std::vector<int> & succ = mb.successors();
  const std::vector<sva::PTransformd> & Xt = mb.transforms();

  for(std::size_t i = 0; i < joints.size(); ++i)
  {
    const auto pred_index = static_cast<size_t>(pred[i]);
    const auto succ_index = static_cast<size_t>(succ[i]);

    mbc.jointConfig[i] = joints[i].pose(mbc.q[i]);
    mbc.parentToSon[i] = mbc.jointConfig[i] * Xt[i];
    mbc.motionSubspace[i] = joints[i].motionSubspace();

    if(pred[i] != -1)
      mbc.bodyPosW[succ_index] = mbc.parentToSon[i] * mbc.bodyPosW[pred_index];
    else
      mbc.bodyPosW[succ_index] = mbc.parentToSon[i];
  }
}

void sForwardKinematics(const MultiBody & mb, MultiBodyConfig & mbc)
{
  checkMatchQ(mb, mbc);

  checkMatchBodyPos(mb, mbc);
  checkMatchJointConf(mb, mbc);
  checkMatchParentToSon(mb, mbc);

  forwardKinematics(mb, mbc);
}

} // namespace rbd
