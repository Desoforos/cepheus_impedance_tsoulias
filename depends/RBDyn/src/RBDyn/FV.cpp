/*
 * Copyright 2012-2019 CNRS-UM LIRMM, CNRS-AIST JRL
 */

// associated header
#include "RBDyn/FV.h"

// includes
// RBdyn
#include "RBDyn/Joint.h"
#include "RBDyn/MultiBody.h"
#include "RBDyn/MultiBodyConfig.h"

namespace rbd
{

void forwardVelocity(const MultiBody & mb, MultiBodyConfig & mbc)
{
  const std::vector<Joint> & joints = mb.joints();
  const std::vector<int> & pred = mb.predecessors();
  const std::vector<int> & succ = mb.successors();

  for(std::size_t i = 0; i < joints.size(); ++i)
  {
    const auto pred_index = static_cast<size_t>(pred[i]);
    const auto succ_index = static_cast<size_t>(succ[i]);

    const sva::PTransformd & X_p_i = mbc.parentToSon[i];

    mbc.jointVelocity[i] = joints[i].motion(mbc.alpha[i]);

    if(pred[i] != -1)
      mbc.bodyVelB[succ_index] = X_p_i * mbc.bodyVelB[pred_index] + mbc.jointVelocity[i];
    else
      mbc.bodyVelB[succ_index] = mbc.jointVelocity[i];

    sva::PTransformd E_0_i(mbc.bodyPosW[succ_index].rotation());
    mbc.bodyVelW[succ_index] = E_0_i.invMul(mbc.bodyVelB[succ_index]);
  }
}

void sForwardVelocity(const MultiBody & mb, MultiBodyConfig & mbc)
{
  checkMatchAlpha(mb, mbc);
  checkMatchBodyPos(mb, mbc);
  checkMatchJointConf(mb, mbc);
  checkMatchParentToSon(mb, mbc);

  checkMatchBodyVel(mb, mbc);
  checkMatchJointVelocity(mb, mbc);
  checkMatchMotionSubspace(mb, mbc);

  forwardVelocity(mb, mbc);
}

} // namespace rbd
