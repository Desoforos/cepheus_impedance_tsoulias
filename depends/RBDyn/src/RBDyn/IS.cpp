/*
 * Copyright 2012-2019 CNRS-UM LIRMM, CNRS-AIST JRL
 */

#include <iostream>

// associated header
#include "RBDyn/IS.h"

// includes
// sva
#include <SpaceVecAlg/SpaceVecAlg>

// RBDyn
#include "RBDyn/FK.h"
#include "RBDyn/FV.h"
#include "RBDyn/Jacobian.h"
#include "RBDyn/MultiBody.h"
#include "RBDyn/MultiBodyConfig.h"
#include "RBDyn/util.hh"

namespace rbd
{
using namespace Eigen;
typedef Matrix<double, 6, 6> Matrix6d;
typedef Matrix<double, 6, 1> Vector6d;

InverseStatics::InverseStatics(const MultiBody & mb)
: f_(static_cast<size_t>(mb.nrBodies())), df_(static_cast<size_t>(mb.nrBodies())),
  jointTorqueDiff_(static_cast<size_t>(mb.nrJoints())), jacW_(static_cast<size_t>(mb.nrBodies())),
  fullJac_(6, mb.nrDof()), jacobianSizeHasBeenSet_(false)
{
  fullJac_.setZero();
  for(size_t i = 0; i < static_cast<size_t>(mb.nrBodies()); ++i)
  {
    jacW_[i] = Jacobian(mb, mb.body(static_cast<int>(i)).name());
    df_[i] = MatrixXd::Zero(6, mb.nrDof());
  }

  for(size_t i = 0; i < static_cast<size_t>(mb.nrJoints()); ++i)
  {
    jointTorqueDiff_[i].resize(mb.joint(static_cast<int>(i)).dof(), mb.nrDof());
    jointTorqueDiff_[i].setZero();
  }
}

void InverseStatics::setJacobianSize(const MultiBody & mb,
                                     const MultiBodyConfig & mbc,
                                     const std::vector<Eigen::MatrixXd> & jacMomentsAndForces)
{
  const std::vector<Body> & bodies = mb.bodies();

  long nColsWanted = 0;
  for(std::size_t i = 0; i < bodies.size(); ++i)
  {
    if(jacMomentsAndForces[i].cols() > nColsWanted)
    {
      nColsWanted = static_cast<long>(jacMomentsAndForces[i].cols());
    }
  }
  for(std::size_t i = 0; i < bodies.size(); ++i)
  {
    if(nColsWanted > df_[i].cols())
    {
      df_[i].resize(6, nColsWanted);
    }
    df_[i].setZero();
    jointTorqueDiff_[i].resize(mbc.motionSubspace[i].cols(), df_[i].cols());
    jointTorqueDiff_[i].setZero();
  }
  jacobianSizeHasBeenSet_ = true;
}

void InverseStatics::inverseStatics(const MultiBody & mb, MultiBodyConfig & mbc)
{
  const std::vector<Body> & bodies = mb.bodies();
  const std::vector<Joint> & joints = mb.joints();
  const std::vector<int> & pred = mb.predecessors();

  sva::MotionVecd a_0(Vector3d::Zero(), mbc.gravity);

  for(std::size_t i = 0; i < bodies.size(); ++i)
  {
    mbc.bodyAccB[i] = mbc.bodyPosW[i] * a_0;
    f_[i] = bodies[i].inertia() * mbc.bodyAccB[i] - mbc.bodyPosW[i].dualMul(mbc.force[i]);
  }

  for(int i = static_cast<int>(joints.size()) - 1; i >= 0; --i)
  {
    // jointTorque is a vector<vector<double>> thus it is necessary to use
    // Eigen::Map to set a vector of elements at once
    // This is identical to do that:
    //
    //    $for (int j = 0; j < joints[i].dof(); ++j)
    //    $  mbc.jointTorque[i][j] = mbc.motionSubspace[i].col(j).transpose() *
    //    f_[i].vector();
    //
    const auto ui = static_cast<size_t>(i);
    VectorXd::Map(mbc.jointTorque[ui].data(), joints[ui].dof()) = f_[ui].vector().transpose() * mbc.motionSubspace[ui];

    if(pred[ui] != -1) f_[static_cast<size_t>(pred[ui])] += mbc.parentToSon[ui].transMul(f_[ui]);
  }
}

void InverseStatics::computeTorqueJacobianJoint(const MultiBody & mb,
                                                MultiBodyConfig & mbc,
                                                const std::vector<MatrixXd> & jacMomentsAndForces)
{
  assert(jacMomentsAndForces.size() == static_cast<size_t>(mb.nrBodies()));

  auto transMat = [](const sva::PTransformd & T)
  {
    Eigen::Matrix6d res;
    Eigen::Matrix3d Rt = T.rotation().transpose();
    res.block(3, 0, 3, 3).setZero();
    res.block(0, 0, 3, 3) = Rt;
    res.block(0, 3, 3, 3) = vector3ToCrossMatrix(T.translation()) * Rt;
    res.block(3, 3, 3, 3) = Rt;
    return res;
  };

  if(!jacobianSizeHasBeenSet_) setJacobianSize(mb, mbc, jacMomentsAndForces);

  const std::vector<Body> & bodies = mb.bodies();
  const std::vector<int> & pred = mb.predecessors();
  const std::vector<Joint> & joints = mb.joints();

  sva::MotionVecd a_0(Vector3d::Zero(), mbc.gravity);

  Matrix6d M;
  Matrix6d N;

  Vector3d aC = a_0.angular();
  Vector3d aF = a_0.linear();
  Matrix3d hatAF, hatAC;
  hatAF = vector3ToCrossMatrix(aF);
  hatAC = vector3ToCrossMatrix(aC);

  for(std::size_t i = 0; i < bodies.size(); ++i)
  {
    df_[i].setZero();
    M.setZero();
    N.setZero();
    // Complete the previously computed jacobian to a full jacobian
    jacW_[i].fullJacobian(mb, jacW_[i].jacobian(mb, mbc), fullJac_);

    mbc.bodyAccB[i] = mbc.bodyPosW[i] * a_0;

    Matrix3d & RW = mbc.bodyPosW[i].rotation();
    Vector3d & tW = mbc.bodyPosW[i].translation();
    Vector3d & fC = mbc.force[i].couple();
    Vector3d & fF = mbc.force[i].force();
    Matrix3d hatFC, hatFF, hathattaC, hathattfF, hattW;
    hatFF = vector3ToCrossMatrix(fF);
    hatFC = vector3ToCrossMatrix(fC);
    hattW = vector3ToCrossMatrix(tW);
    Vector3d hattWaC = hattW * aC;
    Vector3d hattWfF = hattW * fF;
    hathattaC = vector3ToCrossMatrix(hattWaC);
    hathattfF = vector3ToCrossMatrix(hattWfF);

    f_[i] = bodies[i].inertia() * mbc.bodyAccB[i] - mbc.bodyPosW[i].dualMul(mbc.force[i]);

    M.block(0, 0, 3, 3) = RW * hatAC;
    M.block(3, 0, 3, 3) = RW * (-hathattaC + hatAF);
    M.block(3, 3, 3, 3) = RW * hatAC;

    N.block(0, 0, 3, 3) = RW * (hatFC - hathattfF);
    N.block(0, 3, 3, 3) = RW * hatFF;
    N.block(3, 0, 3, 3) = RW * hatFF;

    df_[i].block(0, 0, fullJac_.rows(), fullJac_.cols()) = (bodies[i].inertia().matrix() * M - N) * fullJac_;

    if(jacMomentsAndForces[i].cols() > 0)
    {
      df_[i] += mbc.bodyPosW[i].dualMatrix() * jacMomentsAndForces[i];
    }
  }

  for(int i = static_cast<int>(joints.size()) - 1; i >= 0; --i)
  {
    const auto ui = static_cast<size_t>(i);
    jointTorqueDiff_[ui] = mbc.motionSubspace[ui].transpose() * df_[ui];

    if(pred[ui] != -1)
    {
      const auto pred_index = static_cast<size_t>(pred[ui]);
      Matrix6d transPtS = transMat(mbc.parentToSon[ui]);

      f_[pred_index] += mbc.parentToSon[ui].transMul(f_[ui]);
      df_[pred_index] += transPtS * df_[ui];

      Matrix3d & R = mbc.jointConfig[ui].rotation();
      Vector3d & t = mbc.jointConfig[ui].translation();
      Vector3d RfC = R.transpose() * f_[ui].couple();
      Vector3d RfF = R.transpose() * f_[ui].force();
      Matrix3d hatRfC = vector3ToCrossMatrix(RfC);
      Matrix3d hatRfF = vector3ToCrossMatrix(RfF);
      Matrix6d MJ;

      MJ.block(0, 0, 3, 3) = hatRfC + RfF * t.transpose() - RfF.dot(t) * Matrix3d::Identity();
      MJ.block(0, 3, 3, 3) = -hatRfF;
      MJ.block(3, 0, 3, 3) = hatRfF;
      MJ.block(3, 3, 3, 3).setZero();

      df_[pred_index].block(0, mb.jointPosInDof(i), 6, joints[ui].dof()) -=
          transMat(mb.transforms()[ui]) * MJ * mbc.motionSubspace[ui];
    }
  }
}

void InverseStatics::computeTorqueJacobianJoint(const MultiBody & mb, MultiBodyConfig & mbc)
{
  std::vector<MatrixXd> jacMandF;
  for(int i = 0; i < mb.nrJoints(); ++i) jacMandF.push_back(MatrixXd(0, 0));

  computeTorqueJacobianJoint(mb, mbc, jacMandF);
}

void printMBC(const MultiBody & mb, const MultiBodyConfig & mbc)
{
  std::cout << "mb.bodies() = " << mb.bodies() << std::endl;
  std::cout << "mb.joints() = " << mb.joints() << std::endl;
  std::cout << "mb.predecessors() = " << mb.predecessors() << std::endl;
  std::cout << "mbc.gravity = " << mbc.gravity << std::endl;
  std::cout << "mbc.parentToSon = \n" << mbc.parentToSon << std::endl;
  std::cout << "mbc.bodyPosW = \n" << mbc.bodyPosW << std::endl;
  std::cout << "mbc.force = \n" << mbc.force << std::endl;
  std::cout << "mbc.motionSubspace = \n" << mbc.motionSubspace << std::endl;
}

void InverseStatics::sInverseStatics(const MultiBody & mb, MultiBodyConfig & mbc)
{
  checkMatchAlphaD(mb, mbc);
  checkMatchForce(mb, mbc);
  checkMatchJointConf(mb, mbc);
  checkMatchJointVelocity(mb, mbc);
  checkMatchBodyPos(mb, mbc);
  checkMatchParentToSon(mb, mbc);
  checkMatchBodyVel(mb, mbc);
  checkMatchMotionSubspace(mb, mbc);

  checkMatchBodyAcc(mb, mbc);
  checkMatchJointTorque(mb, mbc);

  inverseStatics(mb, mbc);
}

} // namespace rbd
