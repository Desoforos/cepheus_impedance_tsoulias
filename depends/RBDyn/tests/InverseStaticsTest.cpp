/*
 * Copyright 2012-2019 CNRS-UM LIRMM, CNRS-AIST JRL
 */

// check memory allocation in some method

// includes
// std
#include <iostream>

// boost
#define BOOST_TEST_MODULE Statics
#include <boost/test/included/unit_test.hpp>
#include <boost/test/output_test_stream.hpp>

// SpaceVecAlg
#include <SpaceVecAlg/SpaceVecAlg>

// RBDyn
#include "RBDyn/Body.h"
#include "RBDyn/FD.h"
#include "RBDyn/FK.h"
#include "RBDyn/FV.h"
#include "RBDyn/IS.h"
#include "RBDyn/Joint.h"
#include "RBDyn/MultiBody.h"
#include "RBDyn/MultiBodyConfig.h"
#include "RBDyn/MultiBodyGraph.h"
#include "RBDyn/util.hh"
#include "fixture.hh"

// arm
#include "XXXarm.h"

namespace rbd
{
using namespace Eigen;
Eigen::IOFormat cleanFmt(2, 0, ", ", "\n", "[", "]");

static constexpr double PI = 3.141592653589793238462643383279502884e+00;

void test(boost::shared_ptr<boost::test_tools::output_test_stream> output,
          rbd::MultiBody & mb,
          rbd::MultiBodyConfig & mbc,
          rbd::InverseStatics & IS,
          Eigen::Vector3d q)
{
  Eigen::MatrixXd jacQ(3, 3);
  Eigen::MatrixXd jacF(3, 24);
  std::vector<Eigen::MatrixXd> jacMomentAndForces(4);
  (*output) << "\n\n\nChange config to " << q.transpose() << std::endl;
  mbc.q[1][0] = q(0);
  mbc.q[2][0] = q(1);
  mbc.q[3][0] = q(2);
  forwardKinematics(mb, mbc);
  forwardVelocity(mb, mbc);
  IS.inverseStatics(mb, mbc);
  IS.computeTorqueJacobianJoint(mb, mbc, jacMomentAndForces);

  (*output) << "========================================" << std::endl;
  (*output) << "Results for mbc.q =" << mbc.q << std::endl;
  (*output) << "mbc.jointTorque =\n" << mbc.jointTorque << std::endl;
  for(auto e : IS.f()) (*output) << "IS.f().vector =\n" << e.vector() << std::endl;
  (*output) << "IS.jointTorqueDiff =\n" << IS.jointTorqueDiff() << std::endl;
}

BOOST_AUTO_TEST_CASE(XXXArmTorqueJacobian)
{
  boost::shared_ptr<boost::test_tools::output_test_stream> output = retrievePattern("InverseStaticsTest");

  rbd::MultiBody mb;
  rbd::MultiBodyConfig mbc;
  rbd::MultiBodyGraph mbg;
  std::tie(mb, mbc, mbg) = makeXXXarm();
  rbd::InverseStatics IS(mb);

  test(output, mb, mbc, IS, Vector3d(0, 0, 0));
  test(output, mb, mbc, IS, Vector3d(rbd::PI, 0, 0));
  test(output, mb, mbc, IS, Vector3d(0, rbd::PI / 2, 0));
  test(output, mb, mbc, IS, Vector3d(0.4, 0.1, 0.2));

  std::cout << output->str() << std::endl;
#if not defined __i386__ && not defined __aarch64__
  BOOST_CHECK(output->match_pattern());
#endif
} // end of namespace rbd
} // namespace rbd
