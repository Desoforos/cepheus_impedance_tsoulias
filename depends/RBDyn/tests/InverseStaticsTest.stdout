


Change config to [3](0,0,0)
========================================
Results for mbc.q =[ Empty vector, [ 0], [ 0], [ 0]]
mbc.jointTorque =
[ Empty vector, [ 0], [ 0], [ 0]]
IS.f().vector =
[6](0,0,0,0,39.24,0)
IS.f().vector =
[6](0,0,0,0,29.43,0)
IS.f().vector =
[6](0,0,0,0,19.62,0)
IS.f().vector =
[6](0,0,0,0,9.81,0)
IS.jointTorqueDiff =
[ [0,3](), [3](-29.43,-9.81,0), [3](-9.81,-9.81,0), [3](0,0,0)]



Change config to [3](3.14159,0,0)
========================================
Results for mbc.q =[ Empty vector, [ 3.14159], [ 0], [ 0]]
mbc.jointTorque =
[ Empty vector, [ -3.60414e-15], [ -1.20138e-15], [ 0]]
IS.f().vector =
[6](-3.60414e-15,0,0,0,39.24,7.88861e-31)
IS.f().vector =
[6](-3.60414e-15,0,0,0,-29.43,-3.60414e-15)
IS.f().vector =
[6](-1.20138e-15,0,0,0,-19.62,-2.40276e-15)
IS.f().vector =
[6](0,0,0,0,-9.81,-1.20138e-15)
IS.jointTorqueDiff =
[ [0,3](), [3](29.43,9.81,0), [3](9.81,9.81,0), [3](0,0,0)]



Change config to [3](0,1.5708,0)
========================================
Results for mbc.q =[ Empty vector, [ 0], [ 1.5708], [ 0]]
mbc.jointTorque =
[ Empty vector, [ -9.81], [ -9.81], [ 0]]
IS.f().vector =
[6](-9.81,0,0,0,39.24,0)
IS.f().vector =
[6](-9.81,0,0,0,29.43,0)
IS.f().vector =
[6](-9.81,0,0,0,1.20138e-15,-19.62)
IS.f().vector =
[6](0,0,0,0,6.00689e-16,-9.81)
IS.jointTorqueDiff =
[ [0,3](), [3](-19.62,0,0), [3](-6.00689e-16,-6.00689e-16,0), [3](0,0,0)]



Change config to [3](0.4,0.1,0.2)
========================================
Results for mbc.q =[ Empty vector, [ 0.4], [ 0.1], [ 0.2]]
mbc.jointTorque =
[ Empty vector, [ -12.3436], [ -4.70316], [ 0]]
IS.f().vector =
[6](-12.3436,0,0,0,39.24,1.77636e-15)
IS.f().vector =
[6](-12.3436,0,0,0,27.1068,-11.4606)
IS.f().vector =
[6](-4.70316,0,0,0,17.2182,-9.40633)
IS.f().vector =
[6](0,0,0,0,7.5031,-6.31978)
IS.jointTorqueDiff =
[ [0,3](), [3](-26.6803,-8.60908,0), [3](-8.60908,-8.60908,0), [3](0,0,0)]
